### PR TITLE
Move map_to_workers to the Architecture#worker

### PR DESCRIPTION
**Pull request by vpereira**
Remove method from controller and move it to the model. 
add test case for that

### CODE REVIEW

File: src/api/app/controllers/webui/main_controller.rb [hunk 1]
```diff
@@ -5,7 +5,7 @@ class Webui::MainController < Webui::WebuiController
 
   def gather_busy
     busy = []
-    archs = Architecture.where(available: 1).pluck(:name).map { |arch| map_to_workers(arch) }.uniq
+    archs = Architecture.where(available: 1).pluck(:name).map(&:worker).uniq
     archs.each do |arch|
       starttime = Time.now.to_i - 168.to_i * 3600
       rel = StatusHistory.where("time >= ? AND \`key\` = ?", starttime, 'building_' + arch)
```
Comment (bgeuken) [inline src/api/app/controllers/webui/main_controller.rb]: Doesn't that mean that we now search for a different key?:thinking:

File: src/api/app/controllers/webui/monitor_controller.rb [hunk 2]
```diff
@@ -90,13 +90,14 @@ def events
     check_ajax
     data = {}
 
-    arch = params[:arch] || ''
+    arch = Architecture.find_by(name: params[:arch]) || Architecture.find_by(name: 'x86_64')
```
Comment (dmarcoux) [inline src/api/app/controllers/webui/monitor_controller.rb]: [nitpick]

I would avoid duplicating `Architecture.find_by(name: ...)` and instead do:
```ruby
arch = Architecture.find_by(name: params.fetch(:arch, 'x86_64'))
```

File: src/api/app/controllers/webui/main_controller.rb [hunk 3]
```diff
@@ -5,10 +5,10 @@ class Webui::MainController < Webui::WebuiController
 
   def gather_busy
     busy = []
-    archs = Architecture.where(available: 1).pluck(:name).map { |arch| map_to_workers(arch) }.uniq
+    archs = Architecture.where(available: 1).map(&:worker).uniq
     archs.each do |arch|
       starttime = Time.now.to_i - 168.to_i * 3600
-      rel = StatusHistory.where("time >= ? AND \`key\` = ?", starttime, 'building_' + arch)
+      rel = StatusHistory.where("time >= ? AND \`key\` = ?", starttime, 'building_' + arch.name)
```
Comment (dmarcoux) [inline src/api/app/controllers/webui/main_controller.rb]: Not really needed, but I like that it's clearer now :+1:

File: src/api/app/controllers/webui/monitor_controller.rb [hunk 4]
```diff
@@ -90,13 +90,14 @@ def events
...
+
     range = params[:range]
     ['waiting', 'blocked', 'squeue_high', 'squeue_med'].each do |prefix|
-      data[prefix] = gethistory(prefix + '_' + arch, range, !discard_cache?).map { |time, value| [time * 1000, value] }
+      data[prefix] = gethistory(prefix + '_' + arch.name, range, !discard_cache?).map { |time, value| [time * 1000, value] }
```
Comment (dmarcoux) [inline src/api/app/controllers/webui/monitor_controller.rb]: Not really needed, but I like that it's clearer now :+1:, however could we use String interpolation?

File: src/api/app/controllers/webui/monitor_controller.rb [hunk 5]
```diff
@@ -90,13 +90,14 @@ def events
...
     end
     ['idle', 'building', 'away', 'down', 'dead'].each do |prefix|
-      data[prefix] = gethistory(prefix + '_' + map_to_workers(arch), range, !discard_cache?).map { |time, value| [time * 1000, value] }
+      data[prefix] = gethistory(prefix + '_' + arch.worker, range, !discard_cache?).map { |time, value| [time * 1000, value] }
```
Comment (dmarcoux) [inline src/api/app/controllers/webui/monitor_controller.rb]: Again interpolate that thing!

File: src/api/spec/models/architecture_spec.rb [hunk 6]
```diff
@@ -3,4 +3,11 @@
 RSpec.describe Architecture do
   it { is_expected.to validate_presence_of(:name) }
   it { is_expected.to validate_uniqueness_of(:name) }
+
+  describe '#worker' do
+    let(:i586) { Architecture.find_by(name: 'i586') }
+
+    it { expect(i586).to respond_to(:worker) }
```
Comment (dmarcoux) [inline src/api/spec/models/architecture_spec.rb]: This line is pointless with the line below. I would remove it

File: src/api/app/controllers/webui/monitor_controller.rb (hunk 4 above)
Comment (vpereira) [reply src/api/app/controllers/webui/monitor_controller.rb]: i think yes. lets see

File: src/api/spec/models/architecture_spec.rb (hunk 6 above)
Comment (vpereira) [reply src/api/spec/models/architecture_spec.rb]: yup, TDD, I will remove it

File: src/api/app/controllers/webui/monitor_controller.rb (hunk 2 above)
Comment (coolo) [reply src/api/app/controllers/webui/monitor_controller.rb]: isn't that a different thing? The old code checks if the architecture in params exists and uses a fallback architecture - while your code only takes a fallback if the param is not there.

Just saying

**Comment by vpereira** [comment]
@coolo, searching for HTTP_CACHE_CONTROL, I couldn't find anywhere but in our project. 

Looking into ```rack```,  https://github.com/rack/rack/blob/8d4c8032c400939bad09d48460c425fee077ee03/lib/rack/etag.rb

I found just ```CACHE_CONTROL``` what do you think?

File: src/api/app/controllers/webui/monitor_controller.rb [hunk 7]
```diff
@@ -126,6 +119,23 @@ def discard_cache?
...
+  def gethistory(key, range, cache = 1)
+    cachekey = "#{key}-#{range}"
+    Rails.cache.delete(cachekey, shared: true) unless cache
+    Rails.cache.fetch(cachekey, expires_in: (range.to_i * 3600) / 150, shared: true) do
+      hash = StatusHistory.history_by_key_and_hours(key, range)
+      hash.sort_by { |a| a[0] }
+    end
+  end
+
+  def set_default_arch
```
Comment (dmarcoux) [inline src/api/app/controllers/webui/monitor_controller.rb]: [nitpick]

It's default arch or default architecture? I would follow the same naming.

File: src/api/app/controllers/webui/monitor_controller.rb [hunk 8]
```diff
@@ -126,6 +119,23 @@ def discard_cache?
 
   private
 
+  def get_history(prefix, arch_name, range)
+    gethistory("#{prefix}_#{arch_name}", range, !discard_cache?)
```
Comment (dmarcoux) [inline src/api/app/controllers/webui/monitor_controller.rb]: This is why I miss method overloading... I would get rid of this method and just call this line directly.

**Comment by coolo** [comment]
I don't know. I guess this is confusing request.env and request.headers

File: src/api/app/controllers/webui/monitor_controller.rb [hunk 9]
```diff
@@ -77,40 +76,34 @@ def update_building
...
-      hash.sort_by { |a| a[0] }
-    end
-  end
-
   def events
     check_ajax
     data = {}
 
-    arch = params[:arch] || ''
+    arch = Architecture.find_by(name: params.fetch(:arch, @default_architecture))
```
Comment (dmarcoux) [inline src/api/app/controllers/webui/monitor_controller.rb]: As @coolo pointed out, the behavior is now different, but I think it makes sense to have a default architecture.

**Pull request merged**